src/util-mpm-hs-cache.c

```diff
@@ -244,10 +254,141 @@ void HSSaveCacheIterator(void *data, void *aux)
         return;
     }
 
-    if (HSSaveCache(pd->hs_db, HSHashDb(pd), iter_data->cache_path) == 0) {
+    char hs_db_hash[SC_SHA256_LEN * 2 + 1]; // * 2 for hex +1 for nul terminator
+    if (HSHashDb(pd, hs_db_hash, sizeof(hs_db_hash)) != 0) {
+        return;
+    }
+    if (HSSaveCache(pd->hs_db, hs_db_hash, iter_data->cache_path) == 0) {
         pd->cached = true; // for rule reloads
         iter_data->pd_stats->hs_dbs_cache_saved_cnt++;
     }
 }
 
+/**
+ * \brief Prune the HS cache directory of files older than cutoff time.
+ *
+ * \param dirpath        Path to the cache directory.
+ * \param cutoff         Time cutoff (files older than this will be removed).
+ * \param o_fconsidered  If non-NULL, on return will be set to the number of
+ *                       files considered for pruning.
+ *
+ * \retval Number of files removed on success, negative value on failure.
+ */
+static int32_t HSPruneFilesOlderThan(const char *dirpath, time_t cutoff, int32_t *o_fconsidered)
+{
+    DIR *dir = opendir(dirpath);
+    if (dir == NULL) {
+        return -1;
+    }
+    struct dirent *ent;
+    int32_t removed = 0, considered = 0;
+    char path[PATH_MAX];
+    while ((ent = readdir(dir)) != NULL) {
+        const char *name = ent->d_name;
+        size_t namelen = strlen(name);
+        if (namelen < 3 || strcmp(name + namelen - 3, ".hs") != 0)
```

Comment:
@victorjulien 
To correct myself, the check here scans all Hyperscan cache files.

I am still not fond of the idea of auto-deleting _vN.hs files automatically when _vN+1.hs version comes out. (so deleting v1.hs in our case when v2.hs will be out). Some users may still have one folder for more Suricata instances (it might not even be simultaneous instances, just e.g., Suricata 8/9 for PCAP reading), and on upgrading just some of them, the newer Suricatas would delete the lower version caches.

Now, if users upgrade to 8.0.3, the default pruning is 7 days -- I guess that won't make that much difference when to retain back the storage space, and Suri would be friendlier to more exotic deployments?

Reply:
What are the side effects of deleting more aggressively than expected? Just that start up time is a bit longer the next start up, right? Doesn't seem something we should worry about too much.

Reply:
Yes, the startup times would be longer.
By aggressive removal you would make the cache effectively disabled for older versions.
I see the point of the update encouragement 😄